# Building a spam filter with Naive Bayes algorithm.

In this project, we're going to build a spam filter for SMS messages using the multinomial Naive Bayes algorithm. Our goal is to write a program that classifies new messages with an accuracy greater than 80% — so we expect that more than 80% of the new messages will be classified correctly as spam or ham (non-spam).

To train the algorithm, we'll use a dataset of 5,572 SMS messages that are already classified by humans. The dataset was put together by *Tiago A. Almeida and José María Gómez Hidalgo*, and it can be downloaded from the The [UCI Machine Learning Repository](https://archive.ics.uci.edu/dataset/228/sms+spam+collection).

This project is based on and extends a similar project shared by [DataQuest](https://github.com/dataquestio/solutions/blob/master/Mission433Solutions.ipynb).

We will use **pandas** to manipluate the data and **re** for working with text.

In [2]:
import pandas as pd
import re

## Exploring the Dataset

First we read in the dataset and do basic checks - size, head, tail.

In [4]:
sms_data = pd.read_csv('data/SMSSpamCollection', sep='\t', header=None, names = ['Label', 'SMS'])
print(sms_data.shape)
print( sms_data.head() )
print( sms_data.tail() )

(5572, 2)
  Label                                                SMS
0   ham  Go until jurong point, crazy.. Available only ...
1   ham                      Ok lar... Joking wif u oni...
2  spam  Free entry in 2 a wkly comp to win FA Cup fina...
3   ham  U dun say so early hor... U c already then say...
4   ham  Nah I don't think he goes to usf, he lives aro...
     Label                                                SMS
5567  spam  This is the 2nd time we have tried 2 contact u...
5568   ham               Will ü b going to esplanade fr home?
5569   ham  Pity, * was in mood for that. So...any other s...
5570   ham  The guy did some bitching but I acted like i'd...
5571   ham                         Rofl. Its true to its name


In the head/tail we see that most messages are labelled as ham, some are spam. This is normal. Let's check the exact percentages.

In [5]:
sms_data['Label'].value_counts(normalize=True)

Label
ham     0.865937
spam    0.134063
Name: proportion, dtype: float64

We see above that 86.6% messages are ham, and 13.4% are spam. This sample looks representative based on common experience.

# Working with the data
## Training/Test split
Now we split the data into training and test sets (the usual 80/20 split). 80% for the training and 20% for testing

In [10]:
# First randomise the dataset
data_randomised = sms_data.sample(frac = 1, random_state= 1)

# Calculate index for split at 80% of data
split_index = round(len(data_randomised) * 0.8 )

# Training/ test split at the split_index
training_data = data_randomised[ : split_index ].reset_index(drop=True)
test_data = data_randomised[ split_index: ].reset_index(drop=True)

print("Shape of training data is ", training_data.shape, " and the shape of test data is ", test_data.shape)

# Now do a sanity check to see that both sets of data has the ham/spam ratio of the original data.
print ("Training set: \n", training_data['Label'].value_counts(normalize=True) )
print ("Test set: \n", test_data['Label'].value_counts(normalize=True) )

Shape of training data is  (4458, 2)  and the shape of test data is  (1114, 2)
Training set: 
 Label
ham     0.86541
spam    0.13459
Name: proportion, dtype: float64
Test set: 
 Label
ham     0.868043
spam    0.131957
Name: proportion, dtype: float64


The ratios look similar to the original dataset, which his good. We now move on to clean the dataset.

## Data cleaning
In order to work with the algorithm for calculating probabilities, we need to convert all the text messages to see the number of keywords. The below imae captures the essence of what we need to do.
